toolchain/emcc_toolchain_config.bzl

```diff
@@ -0,0 +1,211 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""This file is configuring an Emscripten toolchain to use for WASM.
+It overwrites the tool paths to point to clang and sets the needed flags for different
+types of actions.
+"""
+
+load("@bazel_tools//tools/build_defs/cc:action_names.bzl", "ACTION_NAMES")
+load(
+    "@bazel_tools//tools/cpp:cc_toolchain_config_lib.bzl",
+    "feature",
+    "flag_group",
+    "flag_set",
+    "tool_path",
+    "with_feature_set",
+)
+
+def _impl(ctx):
+    tool_paths = [
+        tool_path(
+            name = "gcc",
+            path = "emcc.sh",
+        ),
+        tool_path(
+            name = "ld",
+            path = "emcc.sh",
+        ),
+        tool_path(
+            name = "ar",
+            path = "emar.sh",
+        ),
+        tool_path(
+            name = "cpp",
+            path = "false.sh",
+        ),
+        tool_path(
+            name = "gcov",
+            path = "false.sh",
+        ),
+        tool_path(
+            name = "nm",
+            path = "NOT_USED",
+        ),
+        tool_path(
+            name = "objdump",
+            path = "false.sh",
+        ),
+        tool_path(
+            name = "strip",
+            path = "NOT_USED",
+        ),
+    ]
+    preprocessor_compile_actions = [
+        ACTION_NAMES.c_compile,
+        ACTION_NAMES.cpp_compile,
+        ACTION_NAMES.linkstamp_compile,
+        ACTION_NAMES.preprocess_assemble,
+        ACTION_NAMES.cpp_header_parsing,
+        ACTION_NAMES.cpp_module_compile,
+        ACTION_NAMES.clif_match,
+    ]
+
+    all_link_actions = [
+        ACTION_NAMES.cpp_link_executable,
+        ACTION_NAMES.cpp_link_dynamic_library,
+        ACTION_NAMES.cpp_link_nodeps_dynamic_library,
+    ]
+
+    all_compile_actions = [
+        ACTION_NAMES.c_compile,
+        ACTION_NAMES.cpp_compile,
+        ACTION_NAMES.linkstamp_compile,
+        ACTION_NAMES.assemble,
+        ACTION_NAMES.preprocess_assemble,
+        ACTION_NAMES.cpp_header_parsing,
+        ACTION_NAMES.cpp_module_compile,
+        ACTION_NAMES.cpp_module_codegen,
+        ACTION_NAMES.clif_match,
+        ACTION_NAMES.lto_backend,
+    ]
+    toolchain_include_directories_feature = feature(
+        name = "toolchain_include_directories",
+        enabled = True,
+        flag_sets = [
+            flag_set(
+                actions = all_compile_actions,
+                flag_groups = [
+                    flag_group(
+                        flags = [
+                            "-isystem",
+                            "external/emsdk/emsdk/upstream/emscripten/system/include/libcxx",
+                            "-isystem",
+                            "external/emsdk/emsdk/upstream/emscripten/system/lib/libcxxabi/include",
+                            "-isystem",
+                            "external/emsdk/emsdk/upstream/emscripten/system/include/compat",
```

Comment:
what is this?

Reply:
This is a toolchain config used by TFJS https://github.com/tensorflow/tfjs/blob/d0988eedf7bee90fb597e86f0231783e14f70565/tfjs-backend-wasm/toolchain/cc_toolchain_config.bzl#L74-L102

If I understand correctly, it currently tries to use `libc` as it is, without compiling it into Wasm (since a module has imports from `libc`).

Reply:
Let me add more colour to the comment :)
Why do we need to include compat?

Reply:
`compat` is required by TensorFlow.
This happens, when I delete it:
```
ERROR: /root/.cache/bazel/_bazel_root/b6bc0828d5b43c7085b650e0698bf388/external/org_tensorflow/tensorflow/lite/kernels/BUILD:310:1: C++ compilation of rule '@org_tensorflow//tensorflow/lite/kerne
ls:cpu_backend_gemm' failed (Exit 1)                                                                                                                                                               
In file included from external/org_tensorflow/tensorflow/lite/kernels/cpu_backend_gemm_eigen.cc:18:                                                                                                
In file included from external/org_tensorflow/tensorflow/lite/kernels/cpu_backend_gemm_eigen.h:21:                                                                                                 
In file included from external/org_tensorflow/tensorflow/lite/kernels/cpu_backend_context.h:22:                                                                                                    
In file included from external/org_tensorflow/tensorflow/lite/experimental/ruy/context.h:25:                                                                                                       
In file included from external/org_tensorflow/tensorflow/lite/experimental/ruy/prepacked_cache.h:19:                                                                                               
In file included from external/emsdk/emsdk/upstream/emscripten/system/include/libcxx/iostream:38:                                                                                                  
In file included from external/emsdk/emsdk/upstream/emscripten/system/include/libcxx/ios:216:                                                                                                      
external/emsdk/emsdk/upstream/emscripten/system/include/libcxx/__locale:413:32: error: use of undeclared identifier '_CTYPE_S'                                                                     
    static const mask space  = _CTYPE_S;                                                                                                                                                           
                               ^                                                                                                                                                                   
external/emsdk/emsdk/upstream/emscripten/system/include/libcxx/__locale:414:32: error: use of undeclared identifier '_CTYPE_R'                                                                     
    static const mask print  = _CTYPE_R;
```